src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,365 @@
+import type { RequestHandler } from 'express';
+
+import * as _ from 'lodash';
+import {
+	captureException,
+	handleHttpErrors,
+} from '../../../infra/error-handling';
+import { sbvrUtils, errors } from '@balena/pinejs';
+import { getIP } from '../../../lib/utils';
+import {
+	Application,
+	Device,
+	Image,
+	ImageInstall,
+	PickDeferred,
+	Release,
+} from '../../../balena-model';
+import { deleteOldImageInstalls, upsertImageInstall } from './state-patch-v2';
+
+const { BadRequestError, UnauthorizedError, InternalRequestError } = errors;
+const { api } = sbvrUtils;
+
+const validPatchFields = [
+	'status',
+	'os_version',
+	'os_variant',
+	'supervisor_version',
+	'provisioning_progress',
+	'provisioning_state',
+	'ip_address',
+	'mac_address',
+	'api_port',
+	'api_secret',
+	'logs_channel',
+	'memory_usage',
+	'memory_total',
+	'storage_block_device',
+	'storage_usage',
+	'storage_total',
+	'cpu_temp',
+	'cpu_usage',
+	'cpu_id',
+	'is_undervolted',
+	'is_online',
+] as const;
+
+/**
+ * These typings should be used as a guide to what should be sent, but cannot be trusted as what actually *is* sent.
+ */
+export type StatePatchV3Body = {
+	[uuid: string]: {
+		name?: string;
+		status?: string;
+		os_version?: string; // TODO: Should these purely come from the os app?
+		os_variant?: string; // TODO: Should these purely come from the os app?
+		supervisor_version?: string; // TODO: Should this purely come from the supervisor app?
+		provisioning_progress?: number | null; // TODO: should this be reported as part of the os app?
+		provisioning_state?: string | null; // TODO: should this be reported as part of the os app?
+		ip_address?: string;
+		mac_address?: string;
+		api_port?: number; // TODO: should this be reported as part of the supervisor app?
+		api_secret?: string; // TODO: should this be reported as part of the supervisor app?
+		logs_channel?: string; // TODO: should this be reported as part of the supervisor app? or should it not be reported anymore at all?
+		memory_usage?: number;
+		memory_total?: number;
+		storage_block_device?: string;
+		storage_usage?: number;
+		storage_total?: number;
+		cpu_temp?: number;
+		cpu_usage?: number;
+		cpu_id?: string;
+		is_undervolted?: boolean;
+		/**
+		 * Used for setting dependent devices as online
+		 */
+		is_online?: boolean;
+		/**
+		 * Used for setting gateway device of dependent devices
+		 */
+		parent_device?: number;
+		apps?: {
+			[uuid: string]: {
+				/**
+				 * We report the overall release uuid the supervisor considers the active one, even though there may be info for multiple releases.
+				 */
+				release_uuid: string;
+				releases?: {
+					[releaseUUID: string]: {
+						services?: {
+							[name: string]: {
+								image: string;
```

Comment:
This image is meant to be the image name right? I'm not sure how reliably we can get this value, as the image could get deleted/untagged even if the container is still running. Can you get the image install just from the release uuid and serviceName?

Reply:
On second thought, this might be fine, since the supervisor can get the image name from the database, if the image is not on the database (e.g. the DB got deleted), the supervisor will recreate the entry, meaning it generally should be able to get the image name for the running service